In [ ]:
cimport pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
ls

anonymized_hubspot_companies.csv  industry_project_data_merge.ipynb
anonymized_hubspot_deals.csv      mappings.json
anonymized_hubspot_tickets.csv    sample_data/


In [ ]:
tickets = pd.read_csv('anonymized_hubspot_tickets.csv')
deals = pd.read_csv('anonymized_hubspot_deals.csv')
companies = pd.read_csv('anonymized_hubspot_companies.csv')

/tmp/ipykernel_30415/2095908849.py:3: DtypeWarning: Columns (24,25,28,32,45) have mixed types. Specify dtype option on import or set low_memory=False.
  companies = pd.read_csv('anonymized_hubspot_companies.csv')


In [ ]:
deals.columns

Index(['Weighted amount', 'Deal Description',
       'Cumulative time in "BANT Deal. Pain ID'ed (Sales Pipeline)" (HH:mm:ss)',
       'Cumulative time in "Opportunity (Sales Pipeline)" (HH:mm:ss)',
       'Days to close', 'Deal Score', 'Close Date',
       'Deal source attribution 2',
       'Cumulative time in "In Trial - Trial in Progress (Sales Pipeline)" (HH:mm:ss)',
       'Contract Start Date',
       'Cumulative time in "Partner Referrals  (Sales Pipeline)" (HH:mm:ss)',
       'Cumulative time in "Closed Lost (Sales Pipeline)" (HH:mm:ss)',
       'Cumulative time in "Deep Dive. PSP Drafted (Sales Pipeline)" (HH:mm:ss)',
       'Pipeline', 'Record ID', 'Forecast category', 'Original Traffic Source',
       'Associated Company', 'Deal owner', 'Annual contract value',
       'Is Closed (numeric)', 'Amount in company currency',
       'Annual recurring revenue', 'Monthly recurring revenue',
       'Deal probability', 'Associated Company (Primary)', 'Is Closed Won',
       'Forecast 

In [ ]:
deals[['Deal Description','Days to close', 'Is Deal Closed?', 'Is Closed Won','Deal Stage']]

,Deal Description,Days to close,Is Deal Closed?,Is Closed Won,Deal Stage
0,40-50 employees,69.0,False,False,Opportunity
1,"3,000 + Agents.",149.0,False,False,Partner Referrals
2,5-6k agents,149.0,False,False,Partner Referrals
3,Furthest Along - 300 agents,149.0,False,False,Partner Referrals
4,BPO through partnership with AmplifAI,120.0,False,False,Deep Dive. PSP Drafted
...,...,...,...,...,...
588,Development of training Syms for the transitio...,131.0,True,True,Closed Won
589,NaN,131.0,True,True,Closed Won
590,New project with Retail Services. Citi L&D gr...,131.0,True,True,Closed Won
591,Seasonal Hurricane Support for New Hires,246.0,True,True,Closed Won


In [ ]:
### column names of the companies dataset
companies.columns

Index(['Parent Company', 'CCaaS', 'Annual Revenue', 'Target Account',
       'Associated Contact', 'Number of Form Submissions', 'Total Agents',
       'Web Technologies', 'Close Date', '# of Agents Total',
       'Number of times contacted', 'Contact with Primary Company',
       'ICP Fit Level', 'Revenue range', '# of Agents Contracted', 'Record ID',
       'Time Zone', 'Primary Industry', 'Number of Pageviews',
       'Primary Company', 'Year Founded', 'ICP', 'Industry group',
       'Segmentation', 'LMS System', 'SymTrain Use Cases',
       'Associated Company', 'BPO', 'SSO Application', 'State/Region',
       'SymTrain Product', 'Contract End Date', 'SSO Implemented?',
       'Consolidated Industry', 'Type', 'Number of Employees', 'BPO Program',
       'Primary Sub-Industry', 'Number of Sessions', 'WFM', 'Country/Region',
       'Industry', 'Create Date', 'Company name', 'Last Modified Date',
       'BPO Program Tier'],
      dtype='object')

In [ ]:
### column names of the ticket dataset
tickets.columns

Index(['Create date', 'Ticket Tags', 'Stage Date - Project Launch',
       'Associated Contact', 'Category', 'Target Launch Date', 'Kickoff Call',
       'Close date', 'Pipeline', 'Stage Date - Project Initiation',
       'Stage Date - Planning Phase', 'Ticket status',
       'Stage Date - Execution', 'Time to first agent email reply (HH:mm:ss)',
       'Last CES survey rating', 'Priority', 'Associated Company',
       'Who will be creating the Syms?', 'Stage Date - Closure Phase',
       'Stage Date - Monitoring and Control Phase',
       'Associated Company (Primary)', 'Response time (HH:mm:ss)', 'Ticket ID',
       'Time to close (HH:mm:ss)', 'Associated Deal', 'Library index approved',
       'Training: General Overview', 'Latest Milestone',
       'Training: Deployment/User Management Training',
       'Requirements for the Trial', 'Training: Sym Building 101',
       'Last modified date', 'Type', 'Trial End Date',
       'Training: Sym Building 201', 'Stage Date - Converted Won',

In [ ]:
set(deals['Deal Stage'])

{"BANT Deal. Pain ID'ed",
 'Closed Lost',
 'Closed Trial',
 'Closed Won',
 'Contract Sent',
 'Deep Dive. PSP Drafted',
 'In Trial - Trial in Progress',
 'Negotiation',
 'Opportunity',
 'Partner Referrals ',
 'Renewals '}

In [ ]:
companies.columns

Index(['Parent Company', 'CCaaS', 'Annual Revenue', 'Target Account',
       'Associated Contact', 'Number of Form Submissions', 'Total Agents',
       'Web Technologies', 'Close Date', '# of Agents Total',
       'Number of times contacted', 'Contact with Primary Company',
       'ICP Fit Level', 'Revenue range', '# of Agents Contracted', 'Record ID',
       'Time Zone', 'Primary Industry', 'Number of Pageviews',
       'Primary Company', 'Year Founded', 'ICP', 'Industry group',
       'Segmentation', 'LMS System', 'SymTrain Use Cases',
       'Associated Company', 'BPO', 'SSO Application', 'State/Region',
       'SymTrain Product', 'Contract End Date', 'SSO Implemented?',
       'Consolidated Industry', 'Type', 'Number of Employees', 'BPO Program',
       'Primary Sub-Industry', 'Number of Sessions', 'WFM', 'Country/Region',
       'Industry', 'Create Date', 'Company name', 'Last Modified Date',
       'BPO Program Tier'],
      dtype='object')

In [ ]:
companies.head()

,Parent Company,CCaaS,Annual Revenue,Target Account,Associated Contact,Number of Form Submissions,Total Agents,Web Technologies,Close Date,# of Agents Total,...,BPO Program,Primary Sub-Industry,Number of Sessions,WFM,Country/Region,Industry,Create Date,Company name,Last Modified Date,BPO Program Tier
0,NaN,NaN,1.000000e+09,NaN,Contact_ef780380,0.0,NaN,Route 53,NaN,NaN,...,NaN,NaN,0.0,NaN,United States,HVAC and plumbing supply,2024-10-30 10:51,Company_4fc73a2a,2025-02-14 13:39,NaN
1,NaN,NaN,NaN,NaN,Contact_93373ba5,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,Pest Control,2024-09-12 18:30,Company_f3f7e884,2025-02-14 13:39,NaN
2,NaN,NaN,5.000000e+07,NaN,Contact_2e8e0993,0.0,NaN,Youtube; App Nexus; Google Tag Manager; Facebo...,NaN,NaN,...,NaN,NaN,0.0,NaN,United States,Pest Control,2024-09-03 10:44,Company_89929bed,2025-02-14 13:39,NaN
3,NaN,Five9,5.000000e+07,NaN,Contact_635e44ed,0.0,NaN,Postmark; Facebook Advertiser; Salesforce; Goo...,NaN,NaN,...,NaN,NaN,0.0,NaN,United States,Pest Control,2024-08-02 11:36,Company_33dbf591,2025-02-14 13:39,NaN
4,NaN,Five9,1.000000e+08,NaN,Contact_8d055096,0.0,NaN,Microsoft Office 365; Google Tag Manager; Reca...,NaN,NaN,...,NaN,NaN,0.0,NaN,United States,Pest Control,2024-08-02 11:36,Company_a3079821,2025-02-14 13:39,NaN


# load the mapping json file

### The mapping json file contains the dictionaries that can be used to map any of the 2 files together

In [ ]:
### read in the mapping file
import json

with open('mappings.json', 'r') as f:
    mappings = json.load(f)

### check the keys
print(f'mapping keys: {mappings.keys()}')

mapping keys: dict_keys(['CompanyToDeals', 'CompanyToTickets', 'ParentChildRelationships', 'TicketToDeal'])


- `CompanyToDeals`: contains the mappings from company ID to Deals ID
- `CompanyToTickets`: contains the mappings from company ID to Tickets ID
- `TicketToDeal`: contains the mappings from Ticket ID to Deal ID

# Map company and deals dataset

### In order to map the company data and the deals data, we need to first create a one to one mapping between the deals ID and the company ID based on mappings['CompanyToDeals']

In [ ]:
# revert the company-deal mapping
# useful because deals DataFrame has deal records, and now for each deal we can attach the matching company ID, then merge with the companies DataFrame.
deal_to_company = {}

for company_id, deal_list in mappings["CompanyToDeals"].items():
    for deal_id in deal_list:
        deal_to_company[deal_id] = company_id
deal_to_company

{'29547238775': '17771181752',
 '9794988112': '9364308545',
 '9817334075': '9366087515',
 '9817334090': '9366136383',
 '9817320072': '9366140751',
 '9817329956': '9366155050',
 '9817302616': '9366059421',
 '9817306699': '10557622578',
 '9817320073': '9366110525',
 '9783091342': '9355871862',
 '9783148315': '9355871862',
 '9784284936': '9355871862',
 '9784357419': '9355871862',
 '9816968255': '9355871862',
 '9816996368': '9355871862',
 '9816996373': '9355871862',
 '9816996374': '9355871862',
 '9817292301': '9355871862',
 '9817306701': '9355871862',
 '9817311102': '9355871862',
 '9817320086': '9355871862',
 '11993267974': '9355871862',
 '12925391644': '9355871862',
 '22424270143': '9355871862',
 '30181727847': '29943445034',
 '33724679941': '21636976812',
 '32118663507': '21413800792',
 '22254233163': '21413988657',
 '22086573062': '21377449362',
 '18875629455': '19683377975',
 '18246568357': '19634937444',
 '33503208981': '18832393786',
 '17891673997': '17734212240',
 '22813103370': '17

### Add a company ID column in the deals dataset for table merging

In [ ]:
### record ID is the unique deal ID
### first convert it into string type to match the datatype in the dictionary
deals['Record ID'] = deals['Record ID'].astype(str)

#### create a company for company id based on the dictionary using mapping function

deals['Company ID'] = deals['Record ID'].map(deal_to_company)
deals

,Weighted amount,Deal Description,"Cumulative time in ""BANT Deal. Pain ID'ed (Sales Pipeline)"" (HH:mm:ss)","Cumulative time in ""Opportunity (Sales Pipeline)"" (HH:mm:ss)",Days to close,Deal Score,Close Date,Deal source attribution 2,"Cumulative time in ""In Trial - Trial in Progress (Sales Pipeline)"" (HH:mm:ss)",Contract Start Date,...,Is Open (numeric),"Cumulative time in ""Renewals (Sales Pipeline)"" (HH:mm:ss)",Create Date,"Cumulative time in ""Contract Sent (Sales Pipeline)"" (HH:mm:ss)",Total contract value,Last Modified Date,"Cumulative time in ""Closed Won (Sales Pipeline)"" (HH:mm:ss)",Deal Stage,Deal Type,Company ID
0,250.0,40-50 employees,NaN,NaN,69.0,33.0,2025-05-01 15:26,Event,NaN,NaN,...,1.0,NaN,2025-02-20 15:27,NaN,NaN,2025-02-20 16:28,NaN,Opportunity,New,16279026432
1,0.0,"3,000 + Agents.",NaN,NaN,149.0,13.0,2025-07-20 09:06,Referral Partner,NaN,NaN,...,1.0,NaN,2025-02-20 08:07,NaN,NaN,2025-02-20 09:08,NaN,Partner Referrals,New,16275369508
2,0.0,5-6k agents,NaN,NaN,149.0,13.0,2025-07-20 09:03,Referral Partner,NaN,NaN,...,1.0,NaN,2025-02-20 08:04,NaN,NaN,2025-02-20 09:05,NaN,Partner Referrals,New,16275369508
3,0.0,Furthest Along - 300 agents,NaN,NaN,149.0,13.0,2025-07-20 08:55,Referral Partner,NaN,NaN,...,1.0,NaN,2025-02-20 07:58,NaN,NaN,2025-02-20 08:59,NaN,Partner Referrals,New,16275369508
4,10000.0,BPO through partnership with AmplifAI,NaN,NaN,120.0,37.0,2025-06-20 15:01,Master Agent,NaN,NaN,...,1.0,NaN,2025-02-19 14:02,NaN,NaN,2025-02-19 15:03,NaN,Deep Dive. PSP Drafted,New,21636976812
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
588,78000.0,Development of training Syms for the transitio...,NaN,NaN,131.0,NaN,2022-12-20 11:59,SymTrain EE name,NaN,2020-03-13,...,0.0,NaN,2022-08-10 16:48,NaN,NaN,2025-02-20 11:21,NaN,Closed Won,New,9355871862
589,216600.0,NaN,NaN,NaN,131.0,NaN,2022-12-20 11:59,SymTrain EE name,NaN,2020-09-01,...,0.0,NaN,2022-08-10 14:49,NaN,NaN,2025-02-20 11:21,NaN,Closed Won,New,9355871862
590,34200.0,New project with Retail Services. Citi L&D gr...,NaN,NaN,131.0,NaN,2022-12-20 11:59,SymTrain EE name,NaN,2021-01-15,...,0.0,NaN,2022-08-10 14:49,NaN,NaN,2025-02-20 11:21,NaN,Closed Won,New,9355871862
591,53240.0,Seasonal Hurricane Support for New Hires,2044:36:26,NaN,246.0,NaN,2023-04-13 12:59,Referral Partner,NaN,NaN,...,0.0,NaN,2022-08-09 17:32,336:18:05,NaN,2025-02-19 10:45,NaN,Closed Won,New,29543474616


### Merge the deals dataset and the companies dataset

commpany id from the deals table: 'Company ID'

company id from the company table: 'Record ID'

In [ ]:
### convert the integer type to string type for Record ID
companies['Record ID'] = companies['Record ID'].astype(str)


deals_merged = pd.merge(left = deals, right = companies, left_on ='Company ID'  ,right_on ='Record ID' , how = 'left')

In [ ]:
deals_merged.columns

Index(['Weighted amount', 'Deal Description',
       'Cumulative time in "BANT Deal. Pain ID'ed (Sales Pipeline)" (HH:mm:ss)',
       'Cumulative time in "Opportunity (Sales Pipeline)" (HH:mm:ss)',
       'Days to close', 'Deal Score', 'Close Date_x',
       'Deal source attribution 2',
       'Cumulative time in "In Trial - Trial in Progress (Sales Pipeline)" (HH:mm:ss)',
       'Contract Start Date',
       'Cumulative time in "Partner Referrals  (Sales Pipeline)" (HH:mm:ss)',
       'Cumulative time in "Closed Lost (Sales Pipeline)" (HH:mm:ss)',
       'Cumulative time in "Deep Dive. PSP Drafted (Sales Pipeline)" (HH:mm:ss)',
       'Pipeline', 'Record ID_x', 'Forecast category',
       'Original Traffic Source', 'Associated Company_x', 'Deal owner',
       'Annual contract value', 'Is Closed (numeric)',
       'Amount in company currency', 'Annual recurring revenue',
       'Monthly recurring revenue', 'Deal probability',
       'Associated Company (Primary)', 'Is Closed Won', 'For

### The deals_merged dataset now contains both the deals inoformation and the company information. You can now go ahead with data preprocessing to handle all missing values, categorical data encoding, feature selection, feature scaling, and etc, and eventually build models to solve a specific problem.

In [ ]:
#### only keep the data where is deal closed is true
df = deals_merged[deals_merged['Is Deal Closed?']==True]

In [ ]:
df.columns

Index(['Weighted amount', 'Deal Description',
       'Cumulative time in "BANT Deal. Pain ID'ed (Sales Pipeline)" (HH:mm:ss)',
       'Cumulative time in "Opportunity (Sales Pipeline)" (HH:mm:ss)',
       'Days to close', 'Deal Score', 'Close Date_x',
       'Deal source attribution 2',
       'Cumulative time in "In Trial - Trial in Progress (Sales Pipeline)" (HH:mm:ss)',
       'Contract Start Date',
       'Cumulative time in "Partner Referrals  (Sales Pipeline)" (HH:mm:ss)',
       'Cumulative time in "Closed Lost (Sales Pipeline)" (HH:mm:ss)',
       'Cumulative time in "Deep Dive. PSP Drafted (Sales Pipeline)" (HH:mm:ss)',
       'Pipeline', 'Record ID_x', 'Forecast category',
       'Original Traffic Source', 'Associated Company_x', 'Deal owner',
       'Annual contract value', 'Is Closed (numeric)',
       'Amount in company currency', 'Annual recurring revenue',
       'Monthly recurring revenue', 'Deal probability',
       'Associated Company (Primary)', 'Is Closed Won', 'For

In [ ]:
### Is Close Won is that we're trying to predict

### True: a successful deal
### false: deal lost
df['Is Closed Won'].value_counts()

,count
Is Closed Won,
False,266
True,158


In [ ]:
### is closed won is the opposite of is closed lost, make sure you have dropped both columns in your X

np.mean(df['Is Closed Won'] != df['Is closed lost'])

np.float64(1.0)

In [ ]:
### check the percentage of missing values, drop the columns if there are too many missing columns
dict(df.isna().mean())

{'Weighted amount': np.float64(0.04716981132075472),
 'Deal Description': np.float64(0.2028301886792453),
 'Cumulative time in "BANT Deal. Pain ID\'ed (Sales Pipeline)" (HH:mm:ss)': np.float64(0.5518867924528302),
 'Cumulative time in "Opportunity (Sales Pipeline)" (HH:mm:ss)': np.float64(0.9599056603773585),
 'Days to close': np.float64(0.0),
 'Deal Score': np.float64(1.0),
 'Close Date_x': np.float64(0.0),
 'Deal source attribution 2': np.float64(0.014150943396226415),
 'Cumulative time in "In Trial - Trial in Progress (Sales Pipeline)" (HH:mm:ss)': np.float64(0.8985849056603774),
 'Contract Start Date': np.float64(0.8372641509433962),
 'Cumulative time in "Partner Referrals  (Sales Pipeline)" (HH:mm:ss)': np.float64(0.9528301886792453),
 'Cumulative time in "Closed Lost (Sales Pipeline)" (HH:mm:ss)': np.float64(0.9952830188679245),
 'Cumulative time in "Deep Dive. PSP Drafted (Sales Pipeline)" (HH:mm:ss)': np.float64(0.9174528301886793),
 'Pipeline': np.float64(0.0),
 'Record ID_x':

In [ ]:
df['Days to close']

,Days to close
7,2.0
16,0.0
18,0.0
23,31.0
24,32.0
...,...
588,131.0
589,131.0
590,131.0
591,246.0


In [ ]:
### categorical data encoding, I would sugget use label encoding even for the nominal data if it comes with too many categories



### Suggested columns to keep for input data X: 'Amount', 'Annual Revenue', 'Deal Type', 'Country/Region', 'Days to close', 'Primary Industry'....

In [ ]:
print("Industry unique:", df['Industry'].nunique(), "| missing:", df['Industry'].isna().mean().round(2))
print("Primary Industry unique:", df['Primary Industry'].nunique(), "| missing:", df['Primary Industry'].isna().mean().round(2))
print("Consolidated Industry unique:", df['Consolidated Industry'].nunique(), "| missing:", df['Consolidated Industry'].isna().mean().round(2))

print("\nIndustry sample:", df['Industry'].value_counts().head(5).to_dict())
print("\nPrimary Industry sample:", df['Primary Industry'].value_counts().head(5).to_dict())
print("\nConsolidated Industry sample:", df['Consolidated Industry'].value_counts().head(5).to_dict())

Industry unique: 60 | missing: 0.07
Primary Industry unique: 21 | missing: 0.25
Consolidated Industry unique: 24 | missing: 0.05

Industry sample: {'Insurance': 49, 'Business Services': 45, 'Healthcare': 22, 'Financial Services': 21, 'Transportation': 21}

Primary Industry sample: {'Business Services': 62, 'Insurance': 50, 'Transportation': 39, 'Finance': 34, 'Software': 27}

Consolidated Industry sample: {'BPO': 68, 'Banking': 53, 'Insurance': 49, 'Healthcare': 45, 'Transportation and Logistics': 41}


## Feature Selection & Industry Consolidation

After merging the deals and companies datasets, we carefully reviewed all available columns before modeling.

### Leakage Columns Dropped
The following columns were dropped because they directly reveal the outcome and would cause data leakage:
- `Is Closed Won` → this is our target variable `y`
- `Is closed lost` → exact opposite of target, perfect leakage
- `Is Deal Closed?` → always True in training data (we filtered to closed deals)
- `Is Closed (numeric)` / `Is Open (numeric)` → redundant outcome indicators
- `Deal Stage` → contains "Closed Won" / "Closed Lost" labels, direct leakage
- `Close Date` → only populated after a deal closes, post-outcome information

### High-Missing Columns Dropped (>70% missing)
Columns with more than 70% missing values were dropped, as imputing them would introduce more noise than signal given our small dataset size (424 rows).

### ID, Text, and Date Columns Dropped
Columns like `Record ID`, `Deal Name`, `Company name`, `Deal owner`, and date columns were dropped because they are identifiers or free-text fields with no predictive value. `Deal owner` was specifically excluded to avoid overfitting to individual sales representatives in an anonymized dataset.

### Industry Column Selection
Three overlapping industry columns were present in the merged dataset:
- `Industry` — 60 unique values, 7% missing (too granular)
- `Primary Industry` — 21 unique values, 25% missing (too many missing)
- `Consolidated Industry` — 24 unique values, only 5% missing (most standardized)
- `Type` was removed after inspecting its relationship with the target. The crosstab showed that values such as `Customer`, `Suspect`, and `Closed Lost` were strongly associated with deal outcome: `Customer` was mostly won, `Suspect` was entirely lost, and `Closed Lost` was mostly lost. Because these values appear to represent account lifecycle/status rather than independent company characteristics, keeping `Type` could introduce proxy leakage.

We kept `Consolidated Industry` as our single industry feature because it has the best balance of granularity and data completeness, and its categories (e.g., BPO, Banking, Insurance) are clearly standardized. The other two industry columns were dropped as redundant.


### Date Feature Engineering
Raw date columns are not directly interpretable by ML models.
`Create Date_x` was decomposed into three derived features:
- `Create_Year` — captures year-over-year trends in deal success
- `Create_Month` — captures seasonality within a year  
- `Create_Quarter` — captures quarterly sales cycle patterns

The original `Create Date_x` column was then dropped.
All other raw date columns (e.g., `Last Modified Date`, `Last Activity Date`)
were dropped as they are either post-outcome or not useful for prediction.


### Redundant Monetary Columns Dropped
Several monetary columns were derived from each other:
- `Weighted amount` = Amount × Deal probability (derived)
- `Amount in company currency` = Amount in a different unit (redundant)
- `Weighted amount in company currency` = derived from both above (redundant)

We kept `Amount`, `Deal probability`, and `Forecast amount` as they
represent distinct signals. The remaining monetary columns were dropped
to reduce multicollinearity, particularly important for Logistic Regression.




### Zero-Variance Columns Dropped
Columns with only one unique value were identified and dropped,
as they carry no information and cannot help the model distinguish
between won and lost deals. This check is run after all other
preprocessing steps to catch any constants introduced during merging.

In [ ]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score, f1_score, roc_auc_score,
    classification_report, confusion_matrix
)

# Target variable
y = df['Is Closed Won'].astype(int)

# Columns that directly reveal the target or are not useful after filtering to closed deals
leakage_cols = [
    'Is Closed Won',
    'Is closed lost',
    'Is Deal Closed?',
    'Is Closed (numeric)',
    'Is Open (numeric)',
    'Deal Stage',
    'Close Date_x',
    'Close Date_y'
]

# ID, name, free-text, rep/contact, and redundant industry columns
id_or_text_cols = [
    'Record ID_x',
    'Record ID_y',
    'Company ID',
    'Deal Name',
    'Deal Description',
    'Company name',
    'Associated Company_x',
    'Associated Company_y',
    'Associated Company (Primary)',
    'Deal owner',
    'Associated Contact',
    'Contact with Primary Company',
    'Create Date_y',
    'Last Modified Date_x',
    'Last Modified Date_y',
    'Last Activity Date',
    'Industry',
    'Primary Industry',
    'Type'
]

# Redundant or high-dimensional columns
# - weighted/company-currency amount fields overlap with Amount
# - Web Technologies is multi-valued and expands into too many sparse dummy columns
redundant_or_sparse_cols = [
    'Weighted amount',
    'Amount in company currency',
    'Weighted amount in company currency',
    'Web Technologies'
]

# Columns with more than 70% missing values
high_missing_cols = [
    col for col in df.columns
    if df[col].isna().mean() > 0.70
]

# Questionable columns are tested separately before final model selection
questionable_cols = [
    'Days to close',
    'Deal probability',
    'Forecast category',
    'Forecast amount',
    'Cumulative time in "BANT Deal. Pain ID\'ed (Sales Pipeline)" (HH:mm:ss)'
]

def make_X(extra_drop_cols=None):
    """
    Build a modeling matrix from df using one consistent cleaning process.
    extra_drop_cols lets us run controlled experiments on questionable columns.
    """
    if extra_drop_cols is None:
        extra_drop_cols = []

    drop_all = list(set(
        leakage_cols +
        high_missing_cols +
        id_or_text_cols +
        redundant_or_sparse_cols +
        extra_drop_cols
    ))

    df_model = df.drop(columns=[c for c in drop_all if c in df.columns]).copy()

    # Convert raw create date into useful date features, then drop original date column
    if 'Create Date_x' in df_model.columns:
        df_model['Create Date_x'] = pd.to_datetime(df_model['Create Date_x'], errors='coerce')

        df_model['Create_Year'] = df_model['Create Date_x'].dt.year
        df_model['Create_Month'] = df_model['Create Date_x'].dt.month
        df_model['Create_Quarter'] = df_model['Create Date_x'].dt.quarter

        df_model = df_model.drop(columns=['Create Date_x'])

    # Drop columns with only one unique non-missing value
    low_variation_cols = [
        col for col in df_model.columns
        if df_model[col].nunique(dropna=True) <= 1
    ]

    df_model = df_model.drop(columns=low_variation_cols, errors='ignore')

    return df_model, y

X, y = make_X()

print("Dropped leakage:", [c for c in leakage_cols if c in df.columns])
print("\nDropped ID/text/redundant columns:", [c for c in id_or_text_cols if c in df.columns])
print("\nDropped redundant/sparse columns:", [c for c in redundant_or_sparse_cols if c in df.columns])
print("\nDropped >70% missing:", [c for c in high_missing_cols if c in df.columns])
print("\nQuestionable columns kept for now:", [c for c in questionable_cols if c in X.columns])
print("\nRemaining columns:", X.columns.tolist())
print("\nX shape:", X.shape)
print("y shape:", y.shape)
print("Duplicate columns:", X.columns[X.columns.duplicated()].tolist())

Dropped leakage: ['Is Closed Won', 'Is closed lost', 'Is Deal Closed?', 'Is Closed (numeric)', 'Is Open (numeric)', 'Deal Stage', 'Close Date_x', 'Close Date_y']

Dropped ID/text/redundant columns: ['Record ID_x', 'Record ID_y', 'Company ID', 'Deal Name', 'Deal Description', 'Company name', 'Associated Company_x', 'Associated Company_y', 'Associated Company (Primary)', 'Deal owner', 'Associated Contact', 'Contact with Primary Company', 'Create Date_y', 'Last Modified Date_x', 'Last Modified Date_y', 'Last Activity Date', 'Industry', 'Primary Industry', 'Type']

Dropped redundant/sparse columns: ['Weighted amount', 'Amount in company currency', 'Weighted amount in company currency', 'Web Technologies']

Dropped >70% missing: ['Cumulative time in "Opportunity (Sales Pipeline)" (HH:mm:ss)', 'Deal Score', 'Cumulative time in "In Trial - Trial in Progress (Sales Pipeline)" (HH:mm:ss)', 'Contract Start Date', 'Cumulative time in "Partner Referrals  (Sales Pipeline)" (HH:mm:ss)', 'Cumulative 

In [ ]:
X.isna().mean().sort_values(ascending=False).head(15)

,0
# of Agents Total,0.688679
Segmentation,0.629717
"Cumulative time in ""BANT Deal. Pain ID'ed (Sales Pipeline)"" (HH:mm:ss)",0.551887
ICP Fit Level,0.438679
Primary Sub-Industry,0.429245
Year Founded,0.101415
State/Region,0.077830
Annual Revenue,0.075472
Time Zone,0.075472
Number of Employees,0.070755


### Web Technologies Dropped

`Web Technologies` is a semicolon-separated field listing each company's technology stack. Multi-hot encoding this column produced over 100+ sparse binary features, which was too many relative to the dataset size of 424 closed deals.

For the baseline model, this column is removed as part of the redundant/high-dimensional column list. In future work, it could be made usable by grouping technologies into broader categories such as CRM, analytics, communication, or hosting tools.

In [ ]:
### preprocessing helper

def build_preprocessor(X_data):
    numeric_cols = X_data.select_dtypes(include=['int64', 'float64']).columns
    categorical_cols = X_data.select_dtypes(include=['object', 'bool']).columns

    low_cardinality_cols = [
        col for col in categorical_cols
        if X_data[col].nunique(dropna=True) <= 10
    ]

    high_cardinality_cols = [
        col for col in categorical_cols
        if X_data[col].nunique(dropna=True) > 10
    ]

    numeric_transformer = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='median')),
        ('scaler', StandardScaler())
    ])

    low_cardinality_transformer = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='constant', fill_value='Unknown')),
        ('onehot', OneHotEncoder(handle_unknown='ignore'))
    ])

    high_cardinality_transformer = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='constant', fill_value='Unknown')),
        ('ordinal', OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1))
    ])

    preprocessor = ColumnTransformer(
        transformers=[
            ('num', numeric_transformer, numeric_cols),
            ('cat_low', low_cardinality_transformer, low_cardinality_cols),
            ('cat_high', high_cardinality_transformer, high_cardinality_cols)
        ]
    )

    return preprocessor

print("Preprocessing helper ready.")

Preprocessing helper ready.


In [ ]:
#### train-test split

X, y = make_X(extra_drop_cols=[])

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("X_train:", X_train.shape)
print("X_test:", X_test.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)

X_train: (339, 27)
X_test: (85, 27)
y_train: (339,)
y_test: (85,)


In [ ]:
#### Logistic Regression model

from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.metrics import classification_report, confusion_matrix

log_model_balanced = Pipeline(steps=[
    ('preprocessor', build_preprocessor(X_train)),
    ('model', LogisticRegression(max_iter=1000, class_weight='balanced'))
])

log_model_balanced.fit(X_train, y_train)

y_pred = log_model_balanced.predict(X_test)
y_prob = log_model_balanced.predict_proba(X_test)[:, 1]

print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))
print("F1:", f1_score(y_test, y_pred))
print("AUC:", roc_auc_score(y_test, y_prob))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 1.0
Precision: 1.0
Recall: 1.0
F1: 1.0
AUC: 1.0

Confusion Matrix:
[[53  0]
 [ 0 32]]

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        53
           1       1.00      1.00      1.00        32

    accuracy                           1.00        85
   macro avg       1.00      1.00      1.00        85
weighted avg       1.00      1.00      1.00        85



During initial modeling, we observed suspiciously high accuracy (100%) on certain feature configurations. We investigated by running a systematic experiment across feature subsets and used crosstab analysis to identify the source. We found that Deal probability, Forecast category, and Forecast amount are set retroactively by HubSpot after a deal closes, making them post-outcome indicators rather than predictive signals. These were confirmed as leakage and removed. The final model was trained on the cleaned feature set.

In [ ]:
#### Logistic Regression experiments on questionable columns

def evaluate_logistic(X_data, y_data, label):
    preprocessor = build_preprocessor(X_data)

    X_train, X_test, y_train, y_test = train_test_split(
        X_data,
        y_data,
        test_size=0.2,
        random_state=42,
        stratify=y_data
    )

    model = Pipeline(steps=[
        ('preprocessor', preprocessor),
        ('model', LogisticRegression(max_iter=5000, class_weight='balanced'))
    ])

    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)
    y_prob = model.predict_proba(X_test)[:, 1]

    return {
        'Experiment': label,
        'Accuracy': accuracy_score(y_test, y_pred),
        'Precision': precision_score(y_test, y_pred),
        'Recall': recall_score(y_test, y_pred),
        'F1': f1_score(y_test, y_pred),
        'AUC': roc_auc_score(y_test, y_prob),
        'Num Features': X_data.shape[1]
    }

questionable_cols = [
    'Days to close',
    'Deal probability',
    'Forecast category',
    'Forecast amount',
    'Cumulative time in "BANT Deal. Pain ID\'ed (Sales Pipeline)" (HH:mm:ss)'
]

experiments = []

# Strict version: drop all questionable columns
X_strict, y_exp = make_X(extra_drop_cols=questionable_cols)
experiments.append(evaluate_logistic(X_strict, y_exp, 'Strict: no questionable columns'))


# Add one questionable column back at a time
for col in questionable_cols:
    extra_drops = [c for c in questionable_cols if c != col]
    X_temp, y_temp = make_X(extra_drop_cols=extra_drops)
    experiments.append(evaluate_logistic(X_temp, y_temp, f'Add {col}'))

# Add all questionable columns
X_all, y_all = make_X(extra_drop_cols=[])
experiments.append(evaluate_logistic(X_all, y_all, 'Add all questionable columns'))

experiment_results = pd.DataFrame(experiments)
experiment_results.sort_values(by='F1', ascending=False)


,Experiment,Accuracy,Precision,Recall,F1,AUC,Num Features
2,Add Deal probability,1.000000,1.000000,1.00000,1.000000,1.000000,23
6,Add all questionable columns,1.000000,1.000000,1.00000,1.000000,1.000000,27
3,Add Forecast category,1.000000,1.000000,1.00000,1.000000,1.000000,23
5,"Add Cumulative time in ""BANT Deal. Pain ID'ed ...",0.682353,0.580645,0.56250,0.571429,0.691627,23
0,Strict: no questionable columns,0.623529,0.500000,0.59375,0.542857,0.658608,22
4,Add Forecast amount,0.623529,0.500000,0.59375,0.542857,0.659198,23
1,Add Days to close,0.623529,0.500000,0.56250,0.529412,0.663325,23


In [ ]:
print("Deal probability vs target:")
print(pd.crosstab(df['Deal probability'], df['Is Closed Won'], margins=True))

print("\nForecast category vs target:")
print(pd.crosstab(df['Forecast category'], df['Is Closed Won'], margins=True))

print("\nForecast amount missingness vs target:")
print(pd.crosstab(df['Forecast amount'].isna(), df['Is Closed Won'], margins=True))

Deal probability vs target:
Is Closed Won     False  True  All
Deal probability                  
0.0                 266     0  266
1.0                   0   158  158
All                 266   158  424

Forecast category vs target:
Is Closed Won      False  True  All
Forecast category                  
Best case              0     1    1
Closed won             0   157  157
Not forecasted       266     0  266
All                  266   158  424

Forecast amount missingness vs target:
Is Closed Won    False  True  All
Forecast amount                  
False              246   158  404
True                20     0   20
All                266   158  424


In [ ]:
#### final Logistic Regression dataset

# Final decision:
# - Drop Deal probability because it perfectly matches the target in the closed-deal data.
# - Drop Forecast category because it nearly/directly encodes the outcome.
# - Drop BANT cumulative time because it is a high-missing pipeline-duration field
#   and did not improve Logistic Regression performance in the experiment.

final_extra_drop_cols = [
    'Days to close',
    'Deal probability',
    'Forecast category',
    'Cumulative time in "BANT Deal. Pain ID\'ed (Sales Pipeline)" (HH:mm:ss)'
]

X_final, y_final = make_X(extra_drop_cols=final_extra_drop_cols)

print("Final X shape:", X_final.shape)
print("Final y shape:", y_final.shape)
print("Duplicate columns:", X_final.columns[X_final.columns.duplicated()].tolist())
print("\nRemaining columns:")
print(X_final.columns.tolist())

Final X shape: (424, 23)
Final y shape: (424,)
Duplicate columns: []

Remaining columns:
['Deal source attribution 2', 'Original Traffic Source', 'Amount', 'Forecast amount', 'Deal Type', 'Annual Revenue', 'Number of Form Submissions', '# of Agents Total', 'Number of times contacted', 'ICP Fit Level', 'Time Zone', 'Number of Pageviews', 'Year Founded', 'Segmentation', 'State/Region', 'Consolidated Industry', 'Number of Employees', 'Primary Sub-Industry', 'Number of Sessions', 'Country/Region', 'Create_Year', 'Create_Month', 'Create_Quarter']


In [ ]:
#### final Logistic Regression model

numeric_cols = X_final.select_dtypes(include=['int64', 'float64']).columns
categorical_cols = X_final.select_dtypes(include=['object', 'bool']).columns

low_cardinality_cols = [col for col in categorical_cols if X_final[col].nunique() <= 10]
high_cardinality_cols = [col for col in categorical_cols if X_final[col].nunique() > 10]

numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

low_cardinality_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='Unknown')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

high_cardinality_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='Unknown')),
    ('ordinal', OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1))
])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_cols),
        ('cat_low', low_cardinality_transformer, low_cardinality_cols),
        ('cat_high', high_cardinality_transformer, high_cardinality_cols)
    ]
)

X_train, X_test, y_train, y_test = train_test_split(
    X_final,
    y_final,
    test_size=0.2,
    random_state=42,
    stratify=y_final
)

log_model_final = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', LogisticRegression(max_iter=5000, class_weight='balanced'))
])

log_model_final.fit(X_train, y_train)

y_pred = log_model_final.predict(X_test)
y_prob = log_model_final.predict_proba(X_test)[:, 1]

print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))
print("F1:", f1_score(y_test, y_pred))
print("AUC:", roc_auc_score(y_test, y_prob))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 0.6235294117647059
Precision: 0.5
Recall: 0.59375
F1: 0.5428571428571428
AUC: 0.6591981132075472

Confusion Matrix:
[[34 19]
 [13 19]]

Classification Report:
              precision    recall  f1-score   support

           0       0.72      0.64      0.68        53
           1       0.50      0.59      0.54        32

    accuracy                           0.62        85
   macro avg       0.61      0.62      0.61        85
weighted avg       0.64      0.62      0.63        85



## Logistic Regression — Summary

The final Logistic Regression model achieved an accuracy of 62.4%, precision of 0.50, recall of 0.59, F1 score of 0.54, and AUC of 0.66. This model serves as a baseline, but its performance dropped after removing stronger leakage-prone and proxy-leakage features.




The confusion matrix shows that Logistic Regression correctly identified 19 out of 32 won deals, but also produced 19 false positives. This suggests that the linear model struggled to separate won and lost deals using the final cleaned feature set. Because of this, Logistic Regression is useful as an interpretable baseline, but it is not the strongest final model.

In [ ]:
## Random Forest

from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

rf_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', RandomForestClassifier(random_state=42, class_weight='balanced'))
])

# Hyperparameter tuning with 5-fold cross-validation
param_grid = {
    'model__n_estimators': [100, 200, 300],
    'model__max_depth': [3, 5, 10, None],
    'model__min_samples_leaf': [1, 5, 10],
    'model__min_samples_split': [2, 5, 10]
}

rf_grid = GridSearchCV(
    rf_pipeline,
    param_grid,
    cv=5,
    scoring='f1',
    n_jobs=-1
)

rf_grid.fit(X_train, y_train)

print("Best parameters:", rf_grid.best_params_)
print("Best CV F1:", round(rf_grid.best_score_, 4))

# Evaluate best model on test set
y_pred_rf = rf_grid.predict(X_test)
y_prob_rf = rf_grid.predict_proba(X_test)[:, 1]

print("\nAccuracy:", accuracy_score(y_test, y_pred_rf))
print("Precision:", precision_score(y_test, y_pred_rf))
print("Recall:", recall_score(y_test, y_pred_rf))
print("F1:", f1_score(y_test, y_pred_rf))
print("AUC:", roc_auc_score(y_test, y_prob_rf))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_rf))

print("\nClassification Report:")
print(classification_report(y_test, y_pred_rf))

Best parameters: {'model__max_depth': 5, 'model__min_samples_leaf': 5, 'model__min_samples_split': 2, 'model__n_estimators': 100}
Best CV F1: 0.7187

Accuracy: 0.7294117647058823
Precision: 0.6097560975609756
Recall: 0.78125
F1: 0.684931506849315
AUC: 0.8071933962264151

Confusion Matrix:
[[37 16]
 [ 7 25]]

Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.70      0.76        53
           1       0.61      0.78      0.68        32

    accuracy                           0.73        85
   macro avg       0.73      0.74      0.72        85
weighted avg       0.75      0.73      0.73        85



## Random Forest — Summary

The Random Forest model was tuned using 5-fold cross-validation with F1 as the scoring metric. The best configuration was `max_depth=5`, `min_samples_leaf=5`, `min_samples_split=2`, and `n_estimators=100`.

On the test set, Random Forest achieved an accuracy of 72.9%, precision of 0.61, recall of 0.78, F1 score of 0.68, and AUC of 0.81. It correctly identified 25 out of 32 actual won deals, making it the strongest model for catching potential wins.

Although Random Forest produces more false positives than a more conservative model, its high recall is valuable in a sales prioritization setting where missing promising opportunities can be costly.

In [ ]:
## AdaBoost

from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

ada_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', AdaBoostClassifier(random_state=42))
])

param_grid = {
    'model__n_estimators': [50, 100, 200],
    'model__learning_rate': [0.01, 0.1, 0.5, 1.0],
}

ada_grid = GridSearchCV(
    ada_pipeline,
    param_grid,
    cv=5,
    scoring='f1',
    n_jobs=-1
)

ada_grid.fit(X_train, y_train)

print("Best parameters:", ada_grid.best_params_)
print("Best CV F1:", round(ada_grid.best_score_, 4))

y_pred_ada = ada_grid.predict(X_test)
y_prob_ada = ada_grid.predict_proba(X_test)[:, 1]

print("\nAccuracy:", accuracy_score(y_test, y_pred_ada))
print("Precision:", precision_score(y_test, y_pred_ada))
print("Recall:", recall_score(y_test, y_pred_ada))
print("F1:", f1_score(y_test, y_pred_ada))
print("AUC:", roc_auc_score(y_test, y_prob_ada))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_ada))

print("\nClassification Report:")
print(classification_report(y_test, y_pred_ada))

Best parameters: {'model__learning_rate': 1.0, 'model__n_estimators': 200}
Best CV F1: 0.6708

Accuracy: 0.7411764705882353
Precision: 0.65625
Recall: 0.65625
F1: 0.65625
AUC: 0.7889150943396226

Confusion Matrix:
[[42 11]
 [11 21]]

Classification Report:
              precision    recall  f1-score   support

           0       0.79      0.79      0.79        53
           1       0.66      0.66      0.66        32

    accuracy                           0.74        85
   macro avg       0.72      0.72      0.72        85
weighted avg       0.74      0.74      0.74        85



## AdaBoost — Summary

The AdaBoost model was tuned using 5-fold cross-validation with F1 as the scoring metric. The best configuration was `learning_rate=1.0` and `n_estimators=200`.

On the test set, AdaBoost achieved an accuracy of 74.1%, precision of 0.66, recall of 0.66, F1 score of 0.66, and AUC of 0.79. It performed more evenly than Logistic Regression, but it did not match Random Forest's recall or F1 score.

AdaBoost is a reasonable alternative model, but it was not selected as the final model because Random Forest identified more actual won deals.

In [ ]:
model_results = pd.DataFrame([
    {
        'Model': 'Logistic Regression',
        'Accuracy': accuracy_score(y_test, y_pred),
        'Precision': precision_score(y_test, y_pred),
        'Recall': recall_score(y_test, y_pred),
        'F1': f1_score(y_test, y_pred),
        'AUC': roc_auc_score(y_test, y_prob)
    },
    {
        'Model': 'Random Forest',
        'Accuracy': accuracy_score(y_test, y_pred_rf),
        'Precision': precision_score(y_test, y_pred_rf),
        'Recall': recall_score(y_test, y_pred_rf),
        'F1': f1_score(y_test, y_pred_rf),
        'AUC': roc_auc_score(y_test, y_prob_rf)
    },
    {
        'Model': 'AdaBoost',
        'Accuracy': accuracy_score(y_test, y_pred_ada),
        'Precision': precision_score(y_test, y_pred_ada),
        'Recall': recall_score(y_test, y_pred_ada),
        'F1': f1_score(y_test, y_pred_ada),
        'AUC': roc_auc_score(y_test, y_prob_ada)
    }
])

model_results.sort_values(by='F1', ascending=False)

,Model,Accuracy,Precision,Recall,F1,AUC
1,Random Forest,0.729412,0.609756,0.78125,0.684932,0.807193
2,AdaBoost,0.741176,0.656250,0.65625,0.656250,0.788915
0,Logistic Regression,0.623529,0.500000,0.59375,0.542857,0.659198


## Model Comparison & Selection

After removing leakage-prone columns and training three classification models, Random Forest performed best overall based on F1 score and recall for won deals.

| Model | Accuracy | Precision | Recall | F1 | AUC |
|---|---:|---:|---:|---:|---:|
| Random Forest | 0.729 | 0.610 | 0.781 | 0.685 | 0.807 |
| AdaBoost | 0.741 | 0.656 | 0.656 | 0.656 | 0.789 |
| Logistic Regression | 0.624 | 0.500 | 0.594 | 0.543 | 0.659 |

Random Forest is selected as the final model because it achieved the highest F1 score and the highest recall for actual won deals. In this business context, missing a potentially successful deal is costly because the sales team may fail to prioritize a high-potential opportunity. Random Forest correctly identified 25 out of 32 actual wins, compared with 21 out of 32 for AdaBoost and 19 out of 32 for Logistic Regression.

AdaBoost achieved the highest accuracy, but Random Forest produced better recall and F1 score, which are more aligned with the business goal. Logistic Regression remains useful as a baseline, but it struggled after removing leakage-prone features.

Feature importances extracted from the Random Forest model are examined
in the following section to identify which company and deal characteristics
are most associated with successful outcomes.

In [ ]:
#### Feature Importance from Random Forest

feature_names = rf_grid.best_estimator_.named_steps['preprocessor'].get_feature_names_out()
importances = rf_grid.best_estimator_.named_steps['model'].feature_importances_

feature_importance = pd.DataFrame({
    'Feature': feature_names,
    'Importance': importances
}).sort_values(by='Importance', ascending=False)

feature_importance.head(15)

,Feature,Importance
6,num__Number of Pageviews,0.141998
1,num__Forecast amount,0.093782
4,num__# of Agents Total,0.089957
9,num__Number of Sessions,0.089636
3,num__Number of Form Submissions,0.082887
0,num__Amount,0.075704
16,cat_low__Deal Type_New,0.073678
5,num__Number of times contacted,0.071768
8,num__Number of Employees,0.041199
7,num__Year Founded,0.036372


## Feature Importance Interpretation

The Random Forest feature importance results suggest that engagement and company/deal size signals are important predictors of deal outcome. The most important features include `Number of Pageviews`, `Forecast amount`, `# of Agents Total`, `Number of Sessions`, `Number of Form Submissions`, `Amount`, and `Number of times contacted`.

These features suggest that both customer engagement and deal/company scale are associated with deal success. However, feature importance should be interpreted as predictive association, not causation.